compiler/p/codegen/PPCOpsDefines.hpp

```diff
@@ -604,7 +631,105 @@ FORMAT_XT_XA_XB_XC,
 // |                 | RA       | RB       |                      |
 // | 0               | 11       | 16       | 21                   |
 // +-----------------+----------+----------+----------------------+
-FORMAT_RA_RB
+FORMAT_RA_RB,
+FORMAT_RA_RB_MEM,
+
+// Format for instructions with an RT field encoding the target register, and RA and D fields
+// encoding an offset-based memory reference:
+//
+// +------+----------+----------+---------------------------------+
+// |      | RT       | RA       | D                               |
+// | 0    | 6        | 11       | 16                              |
+// +------+----------+----------+---------------------------------+
+//
+// Note: This format is also used for addi-like instructions, even though the ISA specifies that
+//       they use an SI field instead of a D field. The reason for this is that these instructions
+//       act more like the memory instructions than other SI instructions: an RA field value of 0
+//       uses the constant 0 instead of the value of gr0. Additionally, these instructions are
+//       commonly used to calculate the effective address that a load/store would have accessed.
+//       This is safe since the 16-bit SI and D fields are encoded in exactly the same way.
+FORMAT_RT_D16_RA,
+
+// Format for instructions with an FRT field encoding the target FP register, and RA and D fields
+// encoding an offset-based memory reference:
+//
+// +------+----------+----------+---------------------------------+
+// |      | FRT      | RA       | D                               |
+// | 0    | 6        | 11       | 16                              |
+// +------+----------+----------+---------------------------------+
+FORMAT_FRT_D16_RA,
+
+// Format for instructions with an RT field encoding the target register, and RA and DS fields
+// encoding an offset-based memory reference:
+//
+// +------+----------+----------+---------------------------+-----+
+// |      | RT       | RA       | DS                        |     |
+// | 0    | 6        | 11       | 16                        | 30  |
+// +------+----------+----------+---------------------------+-----+
+FORMAT_RT_DS_RA,
+
+// Format for instructions with an RS field encoding the source register, and RA and D fields
+// encoding an offset-based memory reference:
+//
+// +------+----------+----------+---------------------------------+
+// |      | RS       | RA       | D                               |
+// | 0    | 6        | 11       | 16                              |
+// +------+----------+----------+---------------------------------+
+FORMAT_RS_D16_RA,
```

Comment:
there is redundant information here.  RS vs RT already is implicated in Trg1Mem or MemSrc1 instruction.  I am wondering if these can be consolidated.

Reply:
While it is theoretically possible to use one format for these, this would preclude any error checking for incorrect use of a store as a `Trg1Mem` instruction or a load as a `MemSrc1` instruction. Consolidating them wouldn't make any sense IMO unless these two instruction classes were interchangeable, which I don't believe they are due to differing register allocation behaviour.

Reply:
from encoding perspective, they are interchangeable indeed. only the specific instruction field name is different. it is of dubious value or even ultimate do-ability for completely code self bug-proofing, but ok to me for the time being for this particular case.